src/mbgl/gl/vertex_array.hpp

```diff
@@ -0,0 +1,58 @@
+#pragma once
+
+#include <mbgl/gl/object.hpp>
+#include <mbgl/gl/attribute.hpp>
+#include <mbgl/gl/state.hpp>
+#include <mbgl/gl/value.hpp>
+
+#include <array>
+#include <memory>
+
+namespace mbgl {
+namespace gl {
+
+class Context;
+
+class VertexArrayState {
+public:
+    VertexArrayState(UniqueVertexArray vertexArray_, Context& context)
+        : vertexArray(std::move(vertexArray_)),
+          bindings(makeBindings(context, std::make_index_sequence<MAX_ATTRIBUTES>())) {
+    }
+
+    void setDirty() {
+        indexBuffer.setDirty();
+        for (auto& binding : bindings) {
+            binding.setDirty();
+        }
+    }
+
+    UniqueVertexArray vertexArray;
+    State<value::BindElementBuffer> indexBuffer;
+
+    using AttributeState = State<value::VertexAttribute, Context&, AttributeLocation>;
+    std::array<AttributeState, MAX_ATTRIBUTES> bindings;
```

Comment:
Do you see a way to avoid duplicating the reference? It seems like it would require refactoring how we pass additional arguments to `Set`  and probably replacing `State::operator=` with `State::set`. If we're going to do that, I'd rather do it in a separate refactor.